documents/entity/model.go

```diff
@@ -0,0 +1,314 @@
+
+	cliententitypb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/entity"
+
+	"github.com/centrifuge/centrifuge-protobufs/documenttypes"
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/coredocument"
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/entity"
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+	"github.com/centrifuge/precise-proofs/proofs"
+	"github.com/centrifuge/precise-proofs/proofs/proto"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+)
+
+const prefix string = "entity"
+
+// tree prefixes for specific to documents use the second byte of a 4 byte slice by convention
+func compactPrefix() []byte { return []byte{0, 3, 0, 0} }
+
+// Entity implements the documents.Model keeps track of entity related fields and state
+type Entity struct {
+	*documents.CoreDocument
+
+	Identity  *identity.DID
+	LegalName string
+	// address
+	Addresses []*entitypb.Address
+	// tax information
+	PaymentDetails []*entitypb.PaymentDetail
+	// Entity contact list
+	Contacts []*entitypb.Contact
+
+	EntitySalts *proofs.Salts
+}
+
+// getClientData returns the client data from the entity model
+func (e *Entity) getClientData() *cliententitypb.EntityData {
+	var didString string
+	if e.Identity != nil {
+		didString = hexutil.Encode(e.Identity[:])
+	}
+
+	return &cliententitypb.EntityData{
+		Identity:       didString,
+		LegalName:      e.LegalName,
+		Addresses:      e.Addresses,
+		PaymentDetails: e.PaymentDetails,
+		Contacts:       e.Contacts,
+	}
+}
+
+// createP2PProtobuf returns centrifuge protobuf specific entityData
+func (e *Entity) createP2PProtobuf() *entitypb.Entity {
+	var didByte []byte
+	if e.Identity != nil {
+		didByte = e.Identity[:]
+	}
+
+	return &entitypb.Entity{
+		Identity:       didByte,
+		LegalName:      e.LegalName,
+		Addresses:      e.Addresses,
+		PaymentDetails: e.PaymentDetails,
+		Contacts:       e.Contacts,
+	}
+}
+
+// InitEntityInput initialize the model based on the received parameters from the rest api call
+func (e *Entity) InitEntityInput(payload *cliententitypb.EntityCreatePayload, self string) error {
+	err := e.initEntityFromData(payload.Data)
+	if err != nil {
+		return err
+	}
+
+	collaborators := append([]string{self}, payload.Collaborators...)
+	cd, err := documents.NewCoreDocumentWithCollaborators(collaborators, compactPrefix())
+	if err != nil {
+		return errors.New("failed to init core document: %v", err)
+	}
+
+	e.CoreDocument = cd
+	return nil
+}
+
+// initEntityFromData initialises entity from entityData
+func (e *Entity) initEntityFromData(data *cliententitypb.EntityData) error {
+
+	if data.Identity != "" {
+		if did, err := identity.NewDIDFromString(data.Identity); err == nil {
+			e.Identity = &did
+		}
+	}
+	e.LegalName = data.LegalName
+	//e.Addresses = data.Addresses //TODO precise proofs not supporting fields yet
```

Comment:
can you add the unsupported fields to precise proofs and let @vimukthi-git know? Once his integration is done, it should be straightforward